tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,344 @@
+    # This only matters for view ops.
+    maybe_composite1_metadata = backend_indices[DispatchKey.CompositeImplicitAutograd].get_kernel(f)
+    maybe_composite2_metadata = backend_indices[DispatchKey.CompositeExplicitAutograd].get_kernel(f)
+    maybe_cpu_metadata = backend_indices[DispatchKey.CPU].get_kernel(f)
+    # Sadly there are some sparse-only view ops.
+    maybe_sparse_cpu_metadata = backend_indices[DispatchKey.SparseCPU].get_kernel(f)
+    maybe_sparse_csr_cpu_metadata = backend_indices[DispatchKey.SparseCsrCPU].get_kernel(f)
+
+    if maybe_composite1_metadata is not None:
+        return maybe_composite1_metadata.kernel
+    if maybe_composite2_metadata is not None:
+        return maybe_composite2_metadata.kernel
+    if maybe_cpu_metadata is not None:
+        return maybe_cpu_metadata.kernel
+    if maybe_sparse_cpu_metadata is not None:
+        return maybe_sparse_cpu_metadata.kernel
+    if maybe_sparse_csr_cpu_metadata is not None:
+        return maybe_sparse_csr_cpu_metadata.kernel
+    return None
+
+
+# Generates RegisterFunctionalization.cpp
+# These provide the kernels that run the functionalization pass, which can be opted into
+# per backend (e.g. XLA or Vulkan), or as a composable transform (functionalize() in functorch).
+@dataclass(frozen=True)
+class Functionalize:
+    # We don't really need a specific BackendIndex.
+    # We just need to be able to call into the in-tree view ops, at::native::{view},
+    # in order to compute accurate strides information.
+    # Right now, I'm doing that by checking if each op has a cpu or composite kernel (which it currently does, for all view ops)
+    backend_indices: Dict[DispatchKey, BackendIndex]
+    target: Union[
+        Literal[Target.REGISTRATION],
+        Literal[Target.DEFINITION],
+        Literal[Target.DECLARATION]
+    ]
+
+    @method_with_native_function
+    def __call__(self, g: Union[NativeFunction, NativeFunctionsGroup]) -> List[str]:
+        fs = [g] if isinstance(g, NativeFunction) else g.functions()
+        group: Optional[NativeFunctionsGroup] = None if isinstance(g, NativeFunction) else g
+        outputs = []
+        for f in fs:
+            native_api_name = hopefully_get_native_api_name(f, self.backend_indices)
+
+            if not is_view_op(f) and not modifies_arguments(f):
+                continue
+            if self.target is Target.REGISTRATION:
+                output = emit_registration(f, native_api_name)
+            elif self.target is Target.DEFINITION:
+                if is_view_op(f) and f.has_composite_implicit_autograd_kernel:
+                    # See Note [Composite view ops in the functionalization pass]
+                    continue
+                output = emit_definition(f, group, native_api_name)
+            elif self.target is Target.DECLARATION:
+                if is_view_op(f) and not f.has_composite_implicit_autograd_kernel:
+                    output = emit_declaration_for_noncomposite_views(f)
```

Comment:
That's fair. I actually don't need `native_api_name` for the definitions either (I used to, but now that I'm using meta tensors to compute strides I'm just invoking the dispatcher instead of calling at::native:: view kernels directly).

I still think it makes sense to keep the definition and registration code in this same class, but I'll move the code for generating inverse declarations into a different class.